### PR TITLE
Deploy IGP

### LINKED ISSUE #674: Deploy InterchainGasPaymaster on Testnet2

### LINKED ISSUE #752: Deploy InterchainGasPaymaster on mainnet
* Deploy the InterchainGasPaymaster proxy and implementation onto mainnet.

Related ticket to do this on testnet2 that should probably be done first: #674

### PR DESCRIPTION
**Pull request by nambrot**
Deploys IGP on testnet and mainnet

Fixes #674 
Fixes #752

### CODE REVIEW

File: typescript/infra/scripts/deploy-igp.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import { readJSON, writeJSON } from '../src/utils/utils';
+
+import {
+  getCoreContractsSdkFilepath,
+  getCoreEnvironmentConfig,
+  getCoreVerificationDirectory,
+  getEnvironment,
+} from './utils';
+
+class IGPDeployer<
```
Comment (yorhodes) [inline typescript/infra/scripts/deploy-igp.ts:22]: damn you made a class 😮

File: typescript/infra/scripts/deploy-igp.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  Chain extends ChainName,
+> extends AbacusCoreInfraDeployer<Chain> {
+  constructor(
+    multiProvider: MultiProvider<Chain>,
+    configMap: ChainMap<Chain, CoreConfig>,
+    public readonly existingContracts: ChainMap<
+      Chain,
+      CoreContracts<Chain, Chain>
+    >,
```
Comment (yorhodes) [inline typescript/infra/scripts/deploy-igp.ts:31]: still not sure I prefer this to the extending the "resumability" approach

File: typescript/infra/scripts/deploy-igp.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    for (const chain of chains) {
+      const contracts = this.existingContracts[chain];
+      const interchainGasPaymaster = await this.deployProxiedContract(
+        chain,
+        'interchainGasPaymaster',
+        [],
+        contracts.upgradeBeaconController.address,
+        [],
+      );
+      contracts.interchainGasPaymaster = interchainGasPaymaster;
```
Comment (yorhodes) [inline typescript/infra/scripts/deploy-igp.ts:47]: the indirection with `contracts` is hard to read
consider directly assigning to `this.existingContracts[chain].interchainGasPaymaster`

Comment (yorhodes) [reply typescript/infra/scripts/deploy-igp.ts:47]: or use `objMap`

File: typescript/infra/scripts/deploy-igp.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  );
+
+  const existingVerifications = readJSON(
+    getCoreVerificationDirectory(environment),
+    'verification.json',
+  );
+
+  const mergedVerificationInputs = objMap(
+    deployer.verificationInputs,
+    (chain, inputs) => [...existingVerifications[chain], ...inputs],
```
Comment (yorhodes) [inline typescript/infra/scripts/deploy-igp.ts:80]: nit
```suggestion
    (chain, inputs) => existingVerifications[chain].concat(inputs),
```

Comment (yorhodes) [reply typescript/infra/scripts/deploy-igp.ts:80]: afaik "merging" doesn't dedupe anything for arrays

File: typescript/sdk/src/deploy/core/AbacusCoreChecker.ts [hunk 5]
```diff
@@ -227,4 +228,24 @@ export class AbacusCoreChecker<
...
+    const contracts = this.app.getContracts(chain);
+    if (contracts.interchainGasPaymaster.addresses) {
+      await this.checkUpgradeBeacon(
+        chain,
+        'InterchainGasPaymaster',
+        contracts.interchainGasPaymaster.addresses,
+      );
+    } else {
+      this.violations.push({
+        type: CoreViolationType.InterchainGasPaymasterNotDeployed,
```
Comment (yorhodes) [inline typescript/sdk/src/deploy/core/AbacusCoreChecker.ts:242]: meh
are we going to create violations like this for all keys?

File: typescript/infra/scripts/deploy-igp.ts (hunk 2 above)
Comment (nambrot) [reply typescript/infra/scripts/deploy-igp.ts:31]: I agree that that is the right move eventually (i.e. idempotent deploys), but for now, lets keep it this way as the alternative seems like a lot of work?

File: typescript/infra/scripts/deploy-igp.ts (hunk 1 above)
Comment (nambrot) [reply typescript/infra/scripts/deploy-igp.ts:22]: Well I mostly wanted to inherit from deployer to get the utilities. I would have preferred a functional approach but we are where we are 📦

File: typescript/infra/scripts/deploy-igp.ts (hunk 3 above)
Comment (nambrot) [reply typescript/infra/scripts/deploy-igp.ts:47]: Liked the former suggestion

File: typescript/infra/scripts/deploy-igp.ts (hunk 4 above)
Comment (nambrot) [reply typescript/infra/scripts/deploy-igp.ts:80]: Do we actually want to dedupe though? IMO we want to keep all inputs around no? I'm not even sure how a dupe would happen though

File: typescript/sdk/src/deploy/core/AbacusCoreChecker.ts (hunk 5 above)
Comment (nambrot) [reply typescript/sdk/src/deploy/core/AbacusCoreChecker.ts:242]: Good point

Comment (nambrot) [reply typescript/sdk/src/deploy/core/AbacusCoreChecker.ts:242]: Made it more generic

**Pull request merged**